autoprecompiles/src/memory_optimizer.rs

```diff
@@ -68,19 +67,16 @@ fn redundant_memory_interactions_indices<T: FieldElement>(
 
         match mem_int.op {
             MemoryOp::Receive => {
-                if let Some((previous_send, existing_values)) = memory_contents.get(&addr) {
+                // If there is an unconsumed send to this address, consume it.
+                // In that case, we can replace both bus interactions with equality constraints
+                // between the data that would have been sent and received.
+                // TODO: This does not check for equivalent expressions of the same address.
+                //       This is not a soundness issue, but we might miss some optimizations.
+                if let Some((previous_send, existing_values)) = memory_contents.remove(&addr) {
                     for (existing, new) in existing_values.iter().zip(mem_int.data.iter()) {
-                        if existing != new {
```

Comment:
I removed this check, just for keeping things as simple as they can be. Adding a `0 = 0` constraint should not do any harm.